File: scripts/extract_weigmann_2016_species.py
#!/usr/bin/env python3
"""
==============================================================================
Extract and Process Weigmann 2016 Species Checklist
==============================================================================

Purpose: Extract Table II from Weigmann (2016) PDF and convert to CSV

Input: Weigmann_2016_Annotated_global_checklist_of_ChondrichthyesTable_II_rotated.pdf
Output: weigmann_2016_species_checklist.csv

Processing steps:
1. Extract table from PDF using tabula-py
2. Fill down hierarchical columns (Subclass, Order, Family)
3. Clean Family column (remove numbers)
4. Split "Depth distribution (m)" into min_depth_m and max_depth_m
5. Convert column names to lowercase with underscores
6. Pivot geographic distribution to binary columns (WIO, EIO, etc.)

Author: Assistant
Date: 2025-10-02

Requirements:
    pip install tabula-py pandas openpyxl

==============================================================================
"""

import pandas as pd
import numpy as np
import re
from pathlib import Path

# Try importing tabula
try:
    import tabula
    TABULA_AVAILABLE = True
except ImportError:
    TABULA_AVAILABLE = False
    print("WARNING: tabula-py not installed. Install with: pip install tabula-py")

# ==============================================================================
# Configuration
# ==============================================================================

DATA_DIR = Path(__file__).parent.parent / "data"
PDF_PATH = DATA_DIR / "Weigmann_2016_Annotated_global_checklist_of_ChondrichthyesTable_II_rotated.pdf"
OUTPUT_CSV = DATA_DIR / "weigmann_2016_species_checklist.csv"
OUTPUT_WIDE_CSV = DATA_DIR / "weigmann_2016_species_checklist_wide.csv"
GEO_LOOKUP_CSV = DATA_DIR / "lookup_geographic_distribution.csv"
RAW_CSV = DATA_DIR / "weigmann_2016_table2_raw.csv"

# Geographic distribution codes
GEO_CODES = ["WIO", "EIO", "SWP", "NWP", "NEP", "SEP", "SWA", "NWA", "NEA", "SEA"]

# ==============================================================================
# STEP 1: Extract Table from PDF
# ==============================================================================

def extract_table_from_pdf():
    """
    Extract Table II from PDF using tabula-py.

    Returns:
        pd.DataFrame: Extracted table data
    """
    if not TABULA_AVAILABLE:
        print("ERROR: tabula-py not available")
        print("Install with: pip install tabula-py")
        print("\nAlternatively, use Tabula GUI:")
        print("1. Download: https://tabula.technology/")
        print("2. Extract Table II manually")
        print(f"3. Save as: {RAW_CSV}")
        return None

    print("Extracting table from PDF using tabula...")

    # Try to extract all tables (Table II spans many pages)
    # This may require multiple attempts with different page ranges

    try:
        # Extract from all pages (may take time)
        tables = tabula.read_pdf(
            str(PDF_PATH),
            pages="all",
            lattice=True,  # Use lattice mode for tables with lines
            multiple_tables=True
        )

        print(f"Extracted {len(tables)} tables from PDF")

        # Find table with expected columns
        # Table II should have: Subclass, Order, Family #, Scientific name, etc.
        expected_cols = ["Subclass", "Order", "Family", "Scientific name"]

        for i, table in enumerate(tables):
            print(f"\nTable {i+1} columns:", list(table.columns))
            if any(col in str(table.columns) for col in expected_cols):
                print(f"Found Table II candidate at index {i}")
                return table

        print("\nWARNING: Could not automatically identify Table II")
        print("Saving all extracted tables for manual review...")

        for i, table in enumerate(tables):
            output_path = DATA_DIR / f"weigmann_table_{i+1}.csv"
            table.to_csv(output_path, index=False)
            print(f"Saved: {output_path}")

        return tables[0] if tables else None

    except Exception as e:
        print(f"ERROR during PDF extraction: {e}")
        return None

# ==============================================================================
# STEP 2: Load and Process Data
# ==============================================================================

def load_raw_csv():
    """Load raw CSV (either from tabula extraction or manual export)"""

    if RAW_CSV.exists():
        print(f"Loading existing raw CSV: {RAW_CSV}")
        return pd.read_csv(RAW_CSV)
    else:
        print("Raw CSV not found. Attempting PDF extraction...")
        df = extract_table_from_pdf()

        if df is not None:
            df.to_csv(RAW_CSV, index=False)
            print(f"Saved raw extraction to: {RAW_CSV}")

        return df

# ==============================================================================
# STEP 3: Fill Down Hierarchical Columns
# ==============================================================================

def fill_down_hierarchical_columns(df):
    """
    Fill down Subclass, Order, and Family columns.

    Args:
        df (pd.DataFrame): Raw species data

    Returns:
        pd.DataFrame: Processed data with filled hierarchical columns
    """
    print("\nFilling down hierarchical columns...")

    # Columns to fill down
    hierarchical_cols = ["Subclass", "Order", "Family #"]

    for col in hierarchical_cols:
        if col in df.columns:
            df[col] = df[col].fillna(method='ffill')
        else:
            print(f"WARNING: Column '{col}' not found")

    return df

# ==============================================================================
# STEP 4: Clean Family Column
# ==============================================================================

def clean_family_column(df):
    """
    Remove numbers from Family column.
    E.g., "Heterodontidae 1" → "Heterodontidae"

    Args:
        df (pd.DataFrame): Species data

    Returns:
        pd.DataFrame: Data with cleaned Family column
    """
    print("Cleaning Family column...")

    if "Family #" in df.columns:
        # Extract family name only (remove trailing numbers)
        df["Family"] = df["Family #"].str.replace(r"\s+\d+$", "", regex=True)
        df = df.drop(columns=["Family #"])

        # Fill down family names
        df["Family"] = df["Family"].fillna(method='ffill')

    return df

# ==============================================================================
# STEP 5: Split Depth Distribution Column
# ==============================================================================

def parse_depth_range(depth_str):
    """
    Parse depth range string into min and max values.

    Examples:
        "0-590" → (0, 590)
        "3->100" → (3, 100)
        ">1000" → (1000, None)
        "Inshore" → (0, 50)

    Args:
        depth_str (str): Depth range string

    Returns:
        tuple: (min_depth, max_depth)
    """
    if pd.isna(depth_str):
        return (None, None)

    depth_str = str(depth_str).strip()

    # Handle "Inshore"
    if "Inshore" in depth_str or "inshore" in depth_str:
        return (0, 50)

    # Handle ">1000" format
    if depth_str.startswith(">"):
        min_depth = int(re.search(r"\d+", depth_str).group())
        return (min_depth, None)

    # Handle "3->100" format
    if "->" in depth_str:
        parts = depth_str.split("->")
        min_depth = int(re.search(r"\d+", parts[0]).group())
        max_depth = int(re.search(r"\d+", parts[1]).group())
        return (min_depth, max_depth)

    # Handle "0-590" format
    if "-" in depth_str:
        parts = depth_str.split("-")
        try:
            min_depth = int(parts[0])
            max_depth = int(parts[1])
            return (min_depth, max_depth)
        except ValueError:
            return (None, None)

    # Single number
    try:
        depth = int(re.search(r"\d+", depth_str).group())
        return (depth, depth)
    except (AttributeError, ValueError):
        return (None, None)

def split_depth_column(df):
    """
    Split 'Depth distribution (m)' into min_depth_m and max_depth_m.

    Args:
        df (pd.DataFrame): Species data

    Returns:
        pd.DataFrame: Data with split depth columns
    """
    print("Splitting depth distribution column...")

    depth_col = "Depth distribution (m)"

    if depth_col in df.columns:
        # Parse depth ranges
        df[["min_depth_m", "max_depth_m"]] = df[depth_col].apply(
            lambda x: pd.Series(parse_depth_range(x))
        )

        # Keep original for reference
        df = df.rename(columns={depth_col: "depth_raw"})

    return df

# ==============================================================================
# STEP 6: Convert Column Names to Lowercase with Underscores
# ==============================================================================

def clean_column_names(df):
    """
    Convert column names to lowercase with underscores.

    Args:
        df (pd.DataFrame): Species data

    Returns:
        pd.DataFrame: Data with cleaned column names
    """
    print("Converting column names to lowercase...")

    df.columns = (
        df.columns
        .str.lower()
        .str.replace(r"\s+", "_", regex=True)
        .str.replace(r"[()]", "", regex=True)
    )

    return df

# ==============================================================================
# STEP 7: Process Geographic Distribution
# ==============================================================================

def add_geographic_binary_columns(df):
    """
    Add binary columns for each geographic distribution code.

    Args:
        df (pd.DataFrame): Species data

    Returns:
        pd.DataFrame: Data with binary geographic columns
    """
    print("Adding binary geographic distribution columns...")

    geo_col = "geographic_distribution"

    if geo_col in df.columns:
        for code in GEO_CODES:
            col_name = f"geo_{code.lower()}"
            df[col_name] = df[geo_col].str.contains(code, na=False)

    return df

# ==============================================================================
# STEP 8: Main Processing Pipeline
# ==============================================================================

def process_species_data():
    """Main processing pipeline"""

    print("=" * 80)
    print("WEIGMANN 2016 SPECIES CHECKLIST EXTRACTION")
    print("=" * 80)

    # Load raw data
    df = load_raw_csv()

    if df is None:
        print("\nERROR: Could not load data")
        print("\nMANUAL EXTRACTION REQUIRED:")
        print("1. Download Tabula: https://tabula.technology/")
        print("2. Open PDF and extract Table II")
        print(f"3. Save as: {RAW_CSV}")
        return

    print(f"\nRaw data: {len(df)} rows, {len(df.columns)} columns")
    print("Columns:", list(df.columns))

    # Process data
    df = fill_down_hierarchical_columns(df)
    df = clean_family_column(df)
    df = split_depth_column(df)
    df = clean_column_names(df)
    df = add_geographic_binary_columns(df)

    # Reorder columns
    cols_ordered = [
        "subclass",
        "order",
        "family",
        "scientific_name",
        "species_authorship",
        "common_name",
        "maximum_size_mm",
        "min_depth_m",
        "max_depth_m",
    ]

    # Add geographic columns
    geo_cols = [f"geo_{code.lower()}" for code in GEO_CODES]
    cols_ordered.extend(geo_cols)

    # Add remaining columns
    remaining_cols = [c for c in df.columns if c not in cols_ordered]
    cols_ordered.extend(remaining_cols)

    # Select available columns
    cols_final = [c for c in cols_ordered if c in df.columns]
    df = df[cols_final]

    # Save processed data
    df.to_csv(OUTPUT_CSV, index=False)
    print(f"\nSaved processed data: {OUTPUT_CSV}")
    print(f"  - Rows: {len(df)}")
    print(f"  - Columns: {len(df.columns)}")

    # Save wide format (same as long for now)
    df.to_csv(OUTPUT_WIDE_CSV, index=False)
    print(f"Saved wide format: {OUTPUT_WIDE_CSV}")

    # Print summary statistics
    print("\n" + "=" * 80)
    print("SUMMARY STATISTICS")
    print("=" * 80)

    print(f"\nTotal species: {len(df)}")

    if "subclass" in df.columns:
        print("\nSpecies by subclass:")
        print(df["subclass"].value_counts())

    if "family" in df.columns:
        print("\nTop 10 families:")
        print(df["family"].value_counts().head(10))

    if any(f"geo_{code.lower()}" in df.columns for code in GEO_CODES):
        print("\nGeographic distribution:")
        geo_summary = {
            code: df[f"geo_{code.lower()}"].sum()
            for code in GEO_CODES
            if f"geo_{code.lower()}" in df.columns
        }
        for code, count in sorted(geo_summary.items(), key=lambda x: x[1], reverse=True):
            print(f"  {code}: {count} species")

    print("\n" + "=" * 80)
    print("PROCESSING COMPLETE")
    print("=" * 80)

# ==============================================================================
# Entry Point
# ==============================================================================

if __name__ == "__main__":
    process_species_data()

File: scripts/extract_weigmann_2016_species.R
#!/usr/bin/env Rscript
# ==============================================================================
# Extract and Process Weigmann 2016 Species Checklist
# ==============================================================================
#
# Purpose: Extract Table II from Weigmann (2016) PDF and convert to CSV
#
# Input: Weigmann_2016_Annotated_global_checklist_of_ChondrichthyesTable_II_rotated.pdf
# Output: weigmann_2016_species_checklist.csv
#
# Processing steps:
# 1. Extract table from PDF using pdftools + tabulizer
# 2. Fill down hierarchical columns (Subclass, Order, Family)
# 3. Clean Family column (remove numbers, e.g., "Heterodontidae 1" → "Heterodontidae")
# 4. Split "Depth distribution (m)" into min_depth_m and max_depth_m
# 5. Convert column names to lowercase with underscores
# 6. Pivot geographic distribution to binary columns (WIO, EIO, etc.)
#
# Author: Assistant
# Date: 2025-10-02
# ==============================================================================

# Load required packages
library(tidyverse)
library(pdftools)

# NOTE: The PDF has a complex rotated table layout that is difficult to parse
# programmatically. Two approaches:
#
# APPROACH 1 (RECOMMENDED): Manual extraction
#   - Use Tabula (https://tabula.technology/) to manually extract Table II
#   - Export as CSV
#   - Then run this script to process the CSV
#
# APPROACH 2: Automated extraction with tabulizer (may require manual cleanup)
#   - Requires Java and rJava
#   - May not work well with rotated PDF layout

# ==============================================================================
# Configuration
# ==============================================================================

pdf_path <- here::here("data", "Weigmann_2016_Annotated_global_checklist_of_ChondrichthyesTable_II_rotated.pdf")
output_csv <- here::here("data", "weigmann_2016_species_checklist.csv")
output_wide_csv <- here::here("data", "weigmann_2016_species_checklist_wide.csv")
geo_lookup_csv <- here::here("data", "lookup_geographic_distribution.csv")

# ==============================================================================
# STEP 1: Extract Table from PDF
# ==============================================================================

cat("==============================================================================\n")
cat("NOTE: Automated PDF table extraction is complex for rotated tables.\n")
cat("==============================================================================\n\n")

cat("RECOMMENDED WORKFLOW:\n")
cat("1. Download Tabula: https://tabula.technology/\n")
cat("2. Open the PDF in Tabula\n")
cat("3. Select Table II (pages vary, start around page 5 of table)\n")
cat("4. Export as CSV\n")
cat("5. Save as: data/weigmann_2016_table2_raw.csv\n")
cat("6. Re-run this script to process the extracted CSV\n\n")

# Check if raw CSV exists
raw_csv <- here::here("data", "weigmann_2016_table2_raw.csv")

if (!file.exists(raw_csv)) {
  cat("ERROR: Raw CSV not found at:", raw_csv, "\n")
  cat("Please extract Table II using Tabula first (see instructions above)\n")

  # Try automated extraction with pdftools as fallback
  cat("\nAttempting automated text extraction (may require manual cleanup)...\n")

  # Extract text with layout
  text <- pdf_text(pdf_path)

  # Find pages containing "Table II"
  table_pages <- which(str_detect(text, "Table II"))

  cat("Table II found on pages:", paste(table_pages, collapse = ", "), "\n")
  cat("Extracted text saved to: data/weigmann_table2_text.txt\n")

  # Save extracted text for manual review
  writeLines(text, here::here("data", "weigmann_table2_text.txt"))

  cat("\nNEXT STEPS:\n")
  cat("1. Review: data/weigmann_table2_text.txt\n")
  cat("2. Use Tabula for cleaner extraction\n")
  cat("3. Save result as: data/weigmann_2016_table2_raw.csv\n")

  quit(status = 1)
}

# ==============================================================================
# STEP 2: Load and Process Raw CSV
# ==============================================================================

cat("Loading raw CSV...\n")
species_raw <- read_csv(raw_csv, show_col_types = FALSE)

cat("Raw data dimensions:", nrow(species_raw), "rows,", ncol(species_raw), "columns\n")
cat("Column names:\n")
print(names(species_raw))

# ==============================================================================
# STEP 3: Fill Down Hierarchical Columns
# ==============================================================================

cat("\nFilling down hierarchical columns (Subclass, Order, Family)...\n")

species_processed <- species_raw %>%
  # Fill down hierarchical taxonomic columns
  fill(Subclass, .direction = "down") %>%
  fill(Order, .direction = "down") %>%
  fill(`Family #`, .direction = "down")

# ==============================================================================
# STEP 4: Clean Family Column
# ==============================================================================

cat("Cleaning Family column (removing numbers)...\n")

species_processed <- species_processed %>%
  mutate(
    # Extract family name only (remove numbers like "Heterodontidae 1" → "Heterodontidae")
    Family = str_replace(`Family #`, "\\s+\\d+$", ""),
    # Remove the original "Family #" column
    `Family #` = NULL
  ) %>%
  # Fill down family name
  fill(Family, .direction = "down")

# ==============================================================================
# STEP 5: Split Depth Distribution Column
# ==============================================================================

cat("Splitting depth distribution into min_depth_m and max_depth_m...\n")

species_processed <- species_processed %>%
  mutate(
    # Extract min and max depth from "Depth distribution (m)" column
    # Handles formats: "0-590", "3->100", "Inshore", "450-1568", ">1000"
    depth_raw = `Depth distribution (m)`,

    # Extract minimum depth
    min_depth_m = case_when(
      str_detect(depth_raw, "Inshore") ~ 0,
      str_detect(depth_raw, "^>") ~ as.numeric(str_extract(depth_raw, "\\d+")),
      TRUE ~ as.numeric(str_extract(depth_raw, "^\\d+"))
    ),

    # Extract maximum depth
    max_depth_m = case_when(
      str_detect(depth_raw, "Inshore") ~ 50,  # Assume inshore = 0-50m
      str_detect(depth_raw, "^>") ~ NA_real_,  # >1000 = unknown max
      str_detect(depth_raw, "->") ~ as.numeric(str_extract(depth_raw, "(?<=>)\\d+")),
      TRUE ~ as.numeric(str_extract(depth_raw, "(?<=-)\\d+$"))
    ),

    # Remove raw depth column
    `Depth distribution (m)` = NULL
  )

# ==============================================================================
# STEP 6: Convert Column Names to Lowercase with Underscores
# ==============================================================================

cat("Converting column names to lowercase with underscores...\n")

species_processed <- species_processed %>%
  rename_with(
    .fn = ~ str_to_lower(.) %>%
      str_replace_all("\\s+", "_") %>%
      str_replace_all("[()]", "")
  )

# ==============================================================================
# STEP 7: Process Geographic Distribution
# ==============================================================================

cat("Processing geographic distribution column...\n")

# Load geographic distribution lookup
geo_lookup <- read_csv(geo_lookup_csv, show_col_types = FALSE)

cat("Geographic distribution codes:", paste(geo_lookup$abbreviation, collapse = ", "), "\n")

# Add individual binary columns for each geographic area
for (abbrev in geo_lookup$abbreviation) {
  col_name <- paste0("geo_", str_to_lower(abbrev))

  species_processed <- species_processed %>%
    mutate(
      !!col_name := str_detect(geographic_distribution, abbrev)
    )
}

# ==============================================================================
# STEP 8: Save Processed CSV (Long Format)
# ==============================================================================

cat("\nSaving processed data (long format)...\n")

species_long <- species_processed %>%
  select(
    # Taxonomic hierarchy
    subclass,
    order,
    family,
    scientific_name,
    species_authorship,

    # Common details
    common_name,
    maximum_size_mm,

    # Depth
    min_depth_m,
    max_depth_m,
    depth_raw,

    # Geographic distribution (original text)
    geographic_distribution,

    # Geographic binary columns
    starts_with("geo_"),

    # Remarks
    remarks
  )

write_csv(species_long, output_csv)
cat("Saved:", output_csv, "\n")
cat("  - Rows:", nrow(species_long), "\n")
cat("  - Columns:", ncol(species_long), "\n")

# ==============================================================================
# STEP 9: Create Wide Format (Geographic Distribution Pivoted)
# ==============================================================================

cat("\nCreating wide format with binary geographic columns...\n")

# Already done in STEP 7 (binary columns added)
# Just rearrange for clarity

species_wide <- species_long %>%
  select(
    # Taxonomic hierarchy
    subclass,
    order,
    family,
    scientific_name,
    species_authorship,
    common_name,
    maximum_size_mm,
    min_depth_m,
    max_depth_m,

    # Geographic binary columns
    geo_wio,
    geo_eio,
    geo_swp,
    geo_nwp,
    geo_nep,
    geo_sep,
    geo_swa,
    geo_nwa,
    geo_nea,
    geo_sea,

    # Keep original text and remarks
    geographic_distribution,
    depth_raw,
    remarks
  )

write_csv(species_wide, output_wide_csv)
cat("Saved:", output_wide_csv, "\n")
cat("  - Rows:", nrow(species_wide), "\n")
cat("  - Columns:", ncol(species_wide), "\n")

# ==============================================================================
# STEP 10: Summary Statistics
# ==============================================================================

cat("\n==============================================================================\n")
cat("SUMMARY STATISTICS\n")
cat("==============================================================================\n\n")

cat("Total species:", nrow(species_wide), "\n\n")

cat("Species by subclass:\n")
species_wide %>%
  count(subclass, sort = TRUE) %>%
  print()

cat("\nTop 10 families by species count:\n")
species_wide %>%
  count(family, sort = TRUE) %>%
  head(10) %>%
  print()

cat("\nGeographic distribution summary (species per area):\n")
species_wide %>%
  summarise(
    WIO = sum(geo_wio, na.rm = TRUE),
    EIO = sum(geo_eio, na.rm = TRUE),
    SWP = sum(geo_swp, na.rm = TRUE),
    NWP = sum(geo_nwp, na.rm = TRUE),
    NEP = sum(geo_nep, na.rm = TRUE),
    SEP = sum(geo_sep, na.rm = TRUE),
    SWA = sum(geo_swa, na.rm = TRUE),
    NWA = sum(geo_nwa, na.rm = TRUE),
    NEA = sum(geo_nea, na.rm = TRUE),
    SEA = sum(geo_sea, na.rm = TRUE)
  ) %>%
  pivot_longer(everything(), names_to = "area", values_to = "species_count") %>%
  arrange(desc(species_count)) %>%
  print()

cat("\nDepth range summary:\n")
species_wide %>%
  summarise(
    min_depth_median = median(min_depth_m, na.rm = TRUE),
    max_depth_median = median(max_depth_m, na.rm = TRUE),
    max_depth_max = max(max_depth_m, na.rm = TRUE),
    species_with_depth = sum(!is.na(min_depth_m))
  ) %>%
  print()

cat("\n==============================================================================\n")
cat("PROCESSING COMPLETE\n")
cat("==============================================================================\n\n")

cat("Output files:\n")
cat("  1. Long format:  ", output_csv, "\n")
cat("  2. Wide format:  ", output_wide_csv, "\n")
cat("  3. Geo lookup:   ", geo_lookup_csv, "\n")

cat("\nNext steps:\n")
cat("  1. Review CSV files for accuracy\n")
cat("  2. Use wide format for SQL species column generation\n")
cat("  3. Create sp_genus_species columns from scientific_name\n")

# ==============================================================================
# END OF SCRIPT
# ==============================================================================
